src/dbus_api/blockdev.rs

```diff
@@ -0,0 +1,285 @@
+        .access(Access::Read)
+        .emits_changed(EmitsChangedSignal::Const)
+        .on_get(get_parent);
+
+    let uuid_property = f.property::<&str, _>("Uuid", ())
+        .access(Access::Read)
+        .emits_changed(EmitsChangedSignal::Const)
+        .on_get(get_uuid);
+
+    let object_name = format!("{}/{}",
+                              STRATIS_BASE_PATH,
+                              dbus_context.get_next_id().to_string());
+
+    let interface_name = format!("{}.{}", STRATIS_BASE_SERVICE, "blockdev");
+
+    let object_path = f.object_path(object_name, Some(OPContext::new(parent, uuid)))
+        .introspectable()
+        .add(f.interface(interface_name, ())
+                 .add_m(set_userid_method)
+                 .add_p(devnode_property)
+                 .add_p(hardware_id_property)
+                 .add_p(initialization_time_property)
+                 .add_p(total_physical_size_property)
+                 .add_p(pool_property)
+                 .add_p(state_property)
+                 .add_p(user_id_property)
+                 .add_p(uuid_property));
+
+    let path = object_path.get_name().to_owned();
+    dbus_context.actions.borrow_mut().push_add(object_path);
+    path
+}
+
+fn set_user_id(m: &MethodInfo<MTFn<TData>, TData>) -> MethodResult {
+    let message: &Message = m.msg;
+    let mut iter = message.iter_init();
+
+    let new_id: &str = get_next_arg(&mut iter, 0)?;
+
+    let dbus_context = m.tree.get_data();
+    let object_path = m.path.get_name();
+    let return_message = message.method_return();
+    let default_return = MessageItem::Bool(false);
+
+    let blockdev_path = m.tree
+        .get(object_path)
+        .expect("implicit argument must be in tree");
+    let blockdev_data = get_data!(blockdev_path; default_return; return_message);
+
+    let pool_path = get_parent!(m; blockdev_data; default_return; return_message);
+    let pool_uuid = &get_data!(pool_path; default_return; return_message).uuid;
+
+    let mut engine = dbus_context.engine.borrow_mut();
+    let pool = get_mut_pool!(engine; pool_uuid; default_return; return_message);
+
+    let blockdev_uuid = &blockdev_data.uuid;
+    let id_changed = {
+        let blockdev =
+            pool.get_mut_blockdev(blockdev_uuid)
+                .ok_or_else(|| {
+                                MethodErr::failed(&format!("no blockdev with uuid {}",
+                                                           &blockdev_uuid))
+                            })?;
+
+        let id_changed = blockdev.user_id() != &Some(new_id.to_owned());
+
+        blockdev
+            .set_user_id(Some(new_id))
+            .map_err(|err| {
+                         MethodErr::failed(&format!("set_user_id failed for object path {}: {}",
+                                                    object_path,
+                                                    err))
+                     })?;
+        id_changed
+    };
+
+    pool.save_state()
```

Comment:
I think the engine should be doing the deciding about whether or not the pool needs to write out metadata.

Reply:
this is a much bigger issue, e.g. see #583. I'm thinking I should leave this and we can replace it when we have something better in place.

Reply:
Ok. Please leave a FIXME to mark the spot.